### ISSUE TITLE
nIn value mismatch - DL4JInvalidInputException

### ISSUE THREAD

**Issue by adonnini** [deeplearning4j/deeplearning4j#9947]
Network execution on Android fails. Please see below for error log.

MY NETWORK CONFIGURATION - DESKTOP SIDE
----------------------------------------------------------------------
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(140)
                .weightInit(WeightInit.XAVIER)
                .updater(new Nesterovs(0.15, 0.9))
                .list()
                .layer(0, new LSTM.Builder().activation(Activation.TANH).nIn(5).nOut(10)
                        .build())
                .layer(1, new RnnOutputLayer.Builder(LossFunctions.LossFunction.MSE)
                        .activation(Activation.IDENTITY).nIn(10).nOut(2).build())
                .build();

INPUT DATASET PROCESSING AND NETWORK EXECUTION - ANDROID SIDE
-------------------------------------------------------------
        tmpData = new double[userLocationsInsideEstimationRadius.size()][5];
        
        double[] flat = ArrayUtil.flattenDoubleArray(tmpData);
        int[] shape = new int[] {userLocationsInsideEstimationRadius.size(), 5, 1};    //Array shape here
        arrData = Nd4j.create(flat,shape,'c');
        
        MultiLayerNetwork restored = ModelSerializer.restoreMultiLayerNetwork(is);
        NormalizerStandardize savedNetworkNormalizer = ModelSerializer.restoreNormalizerFromFile(savedNetworkArchive);
        INDArray results = restored.output(arrData, false);

Execution fails at line
INDArray results = restored.output(arrData, false);

producing the following error log:

04-03 08:44:48.436: W/System.err(3768): java.lang.RuntimeException: org.deeplearning4j.exception.DL4JInvalidInputException: Received input with size(1) = 5 (input array shape = [101272, 5, 1]); input.size(1) must match layer nIn size (nIn = 3)
04-03 08:44:48.436: W/System.err(3768): 	at com.android.contextq.utils.offlinestorage.UserLocationInformationDatabaseService.addRecordToUserLocationInformationDatabase(UserLocationInformationDatabaseService.java:573)
04-03 08:44:48.436: W/System.err(3768): 	at com.android.contextq.utils.offlinestorage.UserLocationInformationDatabaseService.access$400(UserLocationInformationDatabaseService.java:89)
04-03 08:44:48.436: W/System.err(3768): 	at com.android.contextq.utils.offlinestorage.UserLocationInformationDatabaseService$UserLocationInformationDatabaseUpdateRunnable.run(UserLocationInformationDatabaseService.java:677)
04-03 08:44:48.436: W/System.err(3768): 	at java.lang.Thread.run(Thread.java:923)
04-03 08:44:48.436: W/System.err(3768): Caused by: org.deeplearning4j.exception.DL4JInvalidInputException: Received input with size(1) = 5 (input array shape = [101272, 5, 1]); input.size(1) must match layer nIn size (nIn = 3)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.activateHelper(LSTMHelpers.java:171)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.layers.recurrent.LSTM.activateHelper(LSTM.java:145)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.layers.recurrent.LSTM.activate(LSTM.java:110)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.outputOfLayerDetached(MultiLayerNetwork.java:1349)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.output(MultiLayerNetwork.java:2467)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.output(MultiLayerNetwork.java:2430)
04-03 08:44:48.436: W/System.err(3768): 	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.output(MultiLayerNetwork.java:2421)
04-03 08:44:48.436: W/System.err(3768): 	at com.android.contextq.utils.MobileAnalytics.ClusterProcessing.neuralNetworkloadAndRun(ClusterProcessing.java:7929)
04-03 08:44:48.436: W/System.err(3768): 	at com.android.contextq.utils.offlinestorage.UserLocationInformationDatabaseService.addRecordToUserLocationInformationDatabase(UserLocationInformationDatabaseService.java:570)
04-03 08:44:48.436: W/System.err(3768): 	... 3 more

ATTACHMENT - SAVED NETWORK PRODUCED BY  EXECUTION ON DESKTOP SIDE
-----------------------------------------------------------------
[location_next_neural_network.zip](https://github.com/deeplearning4j/deeplearning4j/files/11138415/location_next_neural_network.zip)

ANDROID SIDE - RESTORED NETWORK SUMMARY
---------------------------------------
04-03 08:44:48.428: I/ClusterProcessing(3768):  - neuralNetworkloadAndRun - restored.summary() - 
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================
04-03 08:44:48.428: I/ClusterProcessing(3768): LayerName (LayerType)     nIn,nOut   TotalParams   ParamsShape                 
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================
04-03 08:44:48.428: I/ClusterProcessing(3768): layer0 (LSTM)             3,10       560           W:{3,40}, RW:{10,40}, b:{40}
04-03 08:44:48.428: I/ClusterProcessing(3768): layer1 (RnnOutputLayer)   10,2       22            W:{10,2}, b:{2}             
04-03 08:44:48.428: I/ClusterProcessing(3768): -------------------------------------------------------------------------------
04-03 08:44:48.428: I/ClusterProcessing(3768):             Total Parameters:  582
04-03 08:44:48.428: I/ClusterProcessing(3768):         Trainable Parameters:  582
04-03 08:44:48.428: I/ClusterProcessing(3768):            Frozen Parameters:  0
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================

MY ENVIRONMENT
---------------------------
dl4jVersion = '1.0.0-M2.1'
openblasVersion = '0.3.21-1.5.8'

Android 11
Debian 12

Android Studio Electric Eel 2022.1.1
IntelliJ IDE 2022.3.2

**Comment by agibsonccc** [comment]
@adonnini can you confirm that the input data and the mobile input data are the same shapes? There's no reason we would throw a validation error on mobile that's different from desktop.

**Comment by adonnini** [comment]
I am pretty sure they are. Attached you will find a sample feature file used to test and one used to train, on the desktop side, 
 
[27.csv](https://github.com/deeplearning4j/deeplearning4j/files/11144738/27.csv)
[29.csv](https://github.com/deeplearning4j/deeplearning4j/files/11144742/29.csv)

and sample input dataset values from the Android side:
[inpuDataSetSampleAndroidSide040423-2.txt](https://github.com/deeplearning4j/deeplearning4j/files/11144893/inpuDataSetSampleAndroidSide040423-2.txt)

Please note two things about the sample input dataset content from the Android side:
1) tmpData is defined as
        tmpData = new double[userLocationsInsideEstimationRadius.size()][5];

2) Because of the large number of records in the log file do not always follow the order column 0 to column 4.

**Comment by agibsonccc** [comment]
@agibsonccc  could you just call:
```
DataSet data = itertator.next();
System.out.println(java.util.Arrays.toString(data.getFeatures().shape());
```

and confirm what shape your training data is? There should be zero guessing involved here. The only problem here is your data pipeline. Dl4j's neural network code isn't mysteriously different on android vs desktop there.

If your inference data is different than whatever your iterator is then you need to fix that.

**Comment by adonnini** [comment]
I must be really doing something wrong.
I added the following code on the desktop side:

        log.info(" Prinitng traindata dataset shape");
        while(trainData.hasNext()) {
            DataSet data = trainData.next();
            System.out.println(java.util.Arrays.toString(data.getFeatures().shape()));
        }
        log.info(" Prinitng testdata dataset shape");
        while(testData.hasNext()) {
            DataSet data = testData.next();
            System.out.println(java.util.Arrays.toString(data.getFeatures().shape()));
        }
The two header print statements are printed. There is no shape information. Here is the output:
o.d.e.q.m.r.LocationNextNeuralNetwork -  Prinitng traindata dataset shape
o.d.e.q.m.r.LocationNextNeuralNetwork -  Prinitng testdata dataset shape

**Comment by agibsonccc** [comment]
@adonnini  your iterator doesn't have a next(). You need to do that before you call fit or call reset().

**Comment by adonnini** [comment]
o.d.e.q.m.r.LocationNextNeuralNetwork -  Printing traindata dataset shape
[16, 5, 3]
o.d.e.q.m.r.LocationNextNeuralNetwork -  Printing testdata dataset shape
[16, 5, 60]

**Comment by agibsonccc** [comment]
@adonnini  it looks like your network for training and inference are different. Is there a reason for that? If you print the 2 summaries you can probably confirm that. From the looks of it you're using an old network.

**Comment by adonnini** [comment]
You are right. I have no idea why the restored network still has 3 for nIn. I had noticed that.

I deleted the saved network zip archive from all locations  before saving it again. It did not make a difference


On the Android side:
04-03 08:44:48.428: I/ClusterProcessing(3768):  - neuralNetworkloadAndRun - restored.summary() - 
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================
04-03 08:44:48.428: I/ClusterProcessing(3768): LayerName (LayerType)     nIn,nOut   TotalParams   ParamsShape                 
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================
04-03 08:44:48.428: I/ClusterProcessing(3768): layer0 (LSTM)             3,10       560           W:{3,40}, RW:{10,40}, b:{40}
04-03 08:44:48.428: I/ClusterProcessing(3768): layer1 (RnnOutputLayer)   10,2       22            W:{10,2}, b:{2}             
04-03 08:44:48.428: I/ClusterProcessing(3768): -------------------------------------------------------------------------------
04-03 08:44:48.428: I/ClusterProcessing(3768):             Total Parameters:  582
04-03 08:44:48.428: I/ClusterProcessing(3768):         Trainable Parameters:  582
04-03 08:44:48.428: I/ClusterProcessing(3768):            Frozen Parameters:  0
04-03 08:44:48.428: I/ClusterProcessing(3768): ===============================================================================

On the desktop side:
o.d.e.q.m.r.LocationNextNeuralNetwork -  - net.toStringFull() - 
===============================================================================
LayerName (LayerType)     nIn,nOut   TotalParams   ParamsShape                 
===============================================================================
layer0 (LSTM)             5,10       640           W:{5,40}, RW:{10,40}, b:{40}
layer1 (RnnOutputLayer)   10,2       22            W:{10,2}, b:{2}             
-------------------------------------------------------------------------------
            Total Parameters:  662
        Trainable Parameters:  662
           Frozen Parameters:  0
===============================================================================

**Comment by adonnini** [comment]
I have the saved network zip archive in two locations:
in the raw folder under res
and in my app data folder

**Comment by agibsonccc** [comment]
@adonnini  you'll have to figure that out. It's probably a cache issue or maybe you need to rebuild the apk.  Given that I'm closing this issue.

**Comment by adonnini** [comment]
Thankyou
